cli/azd/pkg/azdo/service_connection.go

```diff
@@ -0,0 +1,165 @@
+	if err != nil {
+		return err
+	}
+
+	endpointResource := "endpoint"
+	endpointAuthorized := true
+	endpointId := endpoint.Id.String()
+	resources := make([]build.DefinitionResourceReference, 1)
+	resources[0] = build.DefinitionResourceReference{
+		Type:       &endpointResource,
+		Authorized: &endpointAuthorized,
+		Id:         &endpointId,
+	}
+
+	authorizeProjectResourcesArgs := build.AuthorizeProjectResourcesArgs{
+		Project:   &projectId,
+		Resources: &resources,
+	}
+
+	_, err = buildClient.AuthorizeProjectResources(ctx, authorizeProjectResourcesArgs)
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// find service connection by name.
+func serviceConnectionExists(ctx context.Context,
+	client *serviceendpoint.Client,
+	projectId *string,
+	serviceConnectionName *string) (bool, error) {
+
+	endpointNames := make([]string, 1)
+	endpointNames[0] = *serviceConnectionName
+	getServiceEndpointsByNamesArgs := serviceendpoint.GetServiceEndpointsByNamesArgs{
+		Project:       projectId,
+		EndpointNames: &endpointNames,
+	}
+
+	serviceEndpoints, err := (*client).GetServiceEndpointsByNames(ctx, getServiceEndpointsByNamesArgs)
+	if err != nil {
+		return false, err
+	}
+
+	for _, endpoint := range *serviceEndpoints {
+		if *endpoint.Name == *serviceConnectionName && *endpoint.IsReady {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// create a new service connection that will be used in the deployment pipeline
+func CreateServiceConnection(
+	ctx context.Context,
+	connection *azuredevops.Connection,
+	projectId string,
+	azdEnvironment environment.Environment,
+	credentials AzureServicePrincipalCredentials,
+	console input.Console) error {
+
+	client, err := serviceendpoint.NewClient(ctx, connection)
+	if err != nil {
+		return err
+	}
+
+	foundServiceConnection, err := serviceConnectionExists(ctx, &client, &projectId, &ServiceConnectionName)
+	if err != nil {
+		return err
+	}
+
+	// if a service connection exists, skip creation.
+	if foundServiceConnection {
+		console.Message(ctx, output.WithWarningFormat("Service Connection %s already exists. Skipping Service Connection Creation", ServiceConnectionName))
+		return nil
+	}
+
+	createServiceEndpointArgs, err := createAzureRMServiceEndPointArgs(ctx, &projectId, credentials)
+	if err != nil {
+		return err
+	}
+	endpoint, err := client.CreateServiceEndpoint(ctx, createServiceEndpointArgs)
+	if err != nil {
+		return err
+	}
+
+	err = authorizeServiceConnectionToAllPipelines(ctx, projectId, endpoint, connection)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// creates input parameter needed to create the azure rm service connection
+func createAzureRMServiceEndPointArgs(
+	ctx context.Context,
+	projectId *string,
+	credentials AzureServicePrincipalCredentials,
+) (serviceendpoint.CreateServiceEndpointArgs, error) {
+	endpointType := "azurerm"
+	endpointOwner := "library"
+	endpointUrl := "https://management.azure.com/"
```

Comment:
Is this overridable?  Does it need to be?

Reply:
Yes it should be . There's a few spots in the az cli package that is hard coded to this endpoint, but to support sovereign clouds, it should be overridable. I'll add an issue so we can address this.